node/src/components/consensus/highway_core/state.rs

```diff
@@ -479,6 +490,37 @@ impl<C: Context> State<C> {
                 .unwrap_or(false)
     }
 
+    /// Updates `self.pings` with the given timestamp.
+    pub(crate) fn add_ping(&mut self, creator: ValidatorIndex, timestamp: Timestamp) {
+        self.pings[creator] = self.pings[creator].max(timestamp);
+    }
+
+    /// Returns `true` if the latest timestamp we have is less than one maximum round length older
+    /// than the given timestamp.
+    ///
+    /// This is to prevent ping spam: If the incoming ping is only slightly newer than a unit or
+    /// ping we have already received, we drop it without forwarding it to our peers.
+    pub(crate) fn has_ping(&self, creator: ValidatorIndex, timestamp: Timestamp) -> bool {
+        let max_round_len = round_len(self.params.max_round_exp());
+        self.pings[creator] + max_round_len > timestamp
+    }
+
+    /// Returns whether the validator is honest and the latest unit or ping is at most
+    /// `PING_TIMEOUT` maximum round lengths old.
+    pub(crate) fn is_online_and_honest(&self, vidx: ValidatorIndex, now: Timestamp) -> bool {
+        let max_round_len = round_len(self.params.max_round_exp());
+        !self.is_faulty(vidx) && self.pings[vidx] + max_round_len * PING_TIMEOUT >= now
```

Comment:
Can you split it up in to `is_honest` and `is_online` parts?

Personally, I find this version more readable: `self.pings[vidx] >= now - max_round_len * PING_TIMEOUT`

Reply:
As discussed, I'll leave the inequality as it is, to avoid underflows in tests.